nixos/modules/services/networking/crab-hole.nix

```
@@ -0,0 +1,230 @@
+                type = lib.types.submodule {
+                  freeformType = settingsFormat.type;
+                };
+              };
+              name_servers = lib.mkOption {
+                description = "Configuration for the NameServers";
+                type = lib.types.listOf (
+                  lib.types.submodule {
+                    options = {
+                      protocol = lib.mkOption {
+                        type = lib.types.enum [
+                          "udp"
+                          "tls"
+                          "https"
+                          "quic"
+                        ];
+                        example = "tls";
+                        default = "tls";
+                        description = "The protocol to use when communicating with the NameServer";
+                      };
+                      socket_addr = lib.mkOption {
+                        type = lib.types.str;
+                        example = "1.1.1.1:853";
+                        description = "Address of the upstream DNS server";
+                      };
+                      tls_dns_name = lib.mkOption {
+                        type = lib.types.str;
+                        example = "1dot1dot1dot1.cloudflare-dns.com";
+                        description = "Domain name of the upstream DNS server";
+                      };
+                      trust_nx_responses = lib.mkEnableOption "Whether to trust `NXDOMAIN` responses from upstream nameservers";
+                    };
+                  }
+                );
+              };
+            };
+          };
+        };
+      };
+
+      configFile = lib.mkOption {
+        type = lib.types.path;
+        description = "The config file of crab-hole. If files are added via url, make sure the service has access to them. Setting this option will override any configuration applied by the settings option.";
+      };
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    # Warning due to DNSSec issue in crab-hole
+    warnings = lib.optional (cfg.settings.upstream.options.validate or false) ''
+      Validate options will ONLY allow DNSSec domains. See https://github.com/LuckyTurtleDev/crab-hole/issues/29
+    '';
+
+    services.crab-hole.configFile = lib.mkDefault (
+      settingsFormat.generate "crab-hole.toml" cfg.settings
+    );
+    environment.etc."crab-hole.toml".source = cfg.configFile;
+
+    systemd.services.crab-hole = {
+      wantedBy = [ "multi-user.target" ];
+      after = [ "network-online.target" ];
+      wants = [ "network-online.target" ];
+      description = "Crab-hole dns server";
+      environment.HOME = "/var/lib/crab-hole";
+      restartTriggers = [ cfg.configFile ];
+      serviceConfig = {
+        Type = "simple";
+        DynamicUser = true;
+        SupplementaryGroups = lib.strings.concatStringsSep " " cfg.supplementaryGroups;
```

I think systemd already concatenates it for you?
```suggestion
        SupplementaryGroups = cfg.supplementaryGroups;
```